pkg/mutation/mutators/assignimage/imageparser.go

```diff
@@ -0,0 +1,114 @@
+package assignimage
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+type image struct {
+	domain string
+	path   string
+	tag    string
+}
+
+var (
+	// domainRegexp defines a schema for a domain component. Primarily, this is to
+	// avoid the possibility of injecting `/` tokens, which could cause part of
+	// the domain component to "leak" over to the location component, ultimately
+	// causing non convergence.
+	domainRegexp = regexp.MustCompile(`^\w[\w.\-_]*(:\d+)?$`)
+
+	// pathRegexp defines a schema for a location component. It follows the convention
+	// specified in the docker distribution reference. The regex  restricts
+	// location-components to start with an alphanumeric character, with following
+	// parts able to be separated by a separator (one period, one or two
+	// underscore and multiple dashes).
+	pathRegexp = regexp.MustCompile(`^[a-z0-9]+(?:(?:(?:[._/]|__|[-]*)[a-z0-9]+)+)?`)
+
+	// tagRegexp defines a schema for a tag component. It must start with `:` or `@`.
+	tagRegexp = regexp.MustCompile(`(^:[\w][\w.-]{0,127}$)|(^@[A-Za-z][A-Za-z0-9]*([-_+.][A-Za-z][A-Za-z0-9]*)*[:][0-9A-Fa-f]{32,}$)`)
+)
+
+func mutateImage(domain, path, tag, mutableImgRef string) string {
+	oldImg := newImage(mutableImgRef)
+	newImg := oldImg.newMutatedImage(domain, path, tag)
+	return newImg.fullRef()
+}
+
+func newImage(imageRef string) image {
+	domain, remainder := splitDomain(imageRef)
+	var pt string
+	tag := ""
+	if tagSep := strings.IndexAny(remainder, ":@"); tagSep > -1 {
+		pt = remainder[:tagSep]
+		tag = remainder[tagSep:]
+	} else {
+		pt = remainder
+	}
+
+	return image{domain: domain, path: pt, tag: tag}
+}
+
+func (img image) newMutatedImage(domain, path, tag string) image {
+	return image{
+		domain: ignoreUnset(img.domain, domain),
+		path:   ignoreUnset(img.path, path),
+		tag:    ignoreUnset(img.tag, tag),
+	}
+}
+
+// ignoreUnset returns `new` if `new` is set, otherwise it returns `old`.
+func ignoreUnset(old, new string) string {
+	if new != "" {
+		return new
+	}
+	return old
+}
+
+func (img image) fullRef() string {
+	domain := img.domain
+	if domain != "" {
+		domain += "/"
+	}
+	return domain + img.path + img.tag
+}
+
+func splitDomain(name string) (domain, remainder string) {
+	i := strings.IndexRune(name, '/')
+	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != "localhost") {
```

Comment:
Do we want to make sure that the port is specified after the colon?

Reply:
We need to literally use the same test code.

I suggest creating a function named `validateDomain()` and calling it both here and in `validateImageParts()` Maybe also add a comment that all domain validation code needs to live in that function in order to maintain convergence

Reply:
Good point. 

As it's setup right now, `splitDomain()` is called when parsing the mutable's image reference (i.e. the string on the actual object we are mutating). We do not want to validate the mutable, so I don't think putting validation in `splitDomain()` is the right move. 

I added a function `validateDomain()` which, along with the regex validation, calls `splitDomain()` to ensure we can parse the field as a domain. `validateDomain()` is called in `validateImageParts()` only.

Reply:
Nice!